src/main/java/org/dcsa/ebl/repository/impl/ShippingInstructionCustomRepositoryImpl.java

```diff
@@ -0,0 +1,257 @@
+        log.debug("populateCarrierBookingReferences");
+
+        return summaries
+                .reduce(
+                        new ArrayList<ShippingInstructionSummaryTO>(),
+                        (list, value) -> { list.add(value); return list; }
+                )
+                .map(shippingInstructions -> findCarrierBookingReferences(
+                        shippingInstructions.stream().map(si -> si.getShippingInstructionID()).collect(Collectors.toList())
+                    ).map(map -> shippingInstructions.stream()
+                        .map(si -> si.withCarrierBookingReferences(map.getOrDefault(si.getShippingInstructionID(), Collections.emptyList())))
+                        .collect(Collectors.toList())
+                ))
+                .flatMap(monolist -> monolist)
+                .flatMapMany(Flux::fromIterable);
+    }
+
+    @Data
+    @RequiredArgsConstructor
+    private static class Pair {
+        private final String shippingInstructionId;
+        private final String carrierBookingReference;
+    }
+
+    private Mono<Map<String, List<String>>> findCarrierBookingReferences(List<String> shippingInstructionIds) {
+        if (shippingInstructionIds.isEmpty()) {
+            log.debug("No shipping instructions found");
+            return Mono.just(Collections.emptyMap());
+        }
+
+        Select query = Select.builder()
+                .select(List.of(ShippingInstructionSpec.id, ShipmentSpec.carrierBookingReference))
+                .from(ShippingInstructionSpec.table, CargoItemSpec.table, ShipmentEquipmentSpec.table, ShipmentSpec.table)
+                .where(
+                        isEqual(ShippingInstructionSpec.id, CargoItemSpec.shippingInstructionId)
+                                .and(isEqual(CargoItemSpec.shipmentEquipmentId, ShipmentEquipmentSpec.id))
+                                .and(isEqual(ShipmentEquipmentSpec.shipmentId, ShipmentSpec.id))
+                                .and(columnIn(ShippingInstructionSpec.id, shippingInstructionIds))
+                ).build();
+
+        RenderContextFactory factory = new RenderContextFactory(r2dbcDialect);
+        SqlRenderer sqlRenderer = SqlRenderer.create(factory.createRenderContext());
+
+        String sql = sqlRenderer.render(query);
+        log.debug("select = {}", sql);
+
+        return addBinds(client.sql(sql), shippingInstructionIds)
+                .map(row ->
+                    new Pair(
+                            row.get("id", String.class),
+                            row.get("carrier_booking_reference", String.class)
+                            )
+                )
+                .all()
+                .reduce(
+                        new HashMap<>(),
+                        (map, value) -> { map.computeIfAbsent(value.shippingInstructionId, k -> new ArrayList<>()).add(value.carrierBookingReference); return map; }
+                );
```

Comment:
I think we should use `.collect(HashMap::new, (map, value) -> { ... })` instead as it better describes the operation (in Flux terms and it avoids the `return` as reactor knows it is building a collection now).

Reply:
Done